contracts/Gap.sol

```diff
@@ -4,34 +4,117 @@ pragma solidity 0.8.19;
+            "Not owner."
+        );
+    }
+
+    ///
+    /// Verify if msg.sender owns the set of attestations
+    ///
+    function canAttestToRefs(AttestationRequestData[] memory datas)
+        private
+        view
+    {
+        for (uint256 j = 0; j < datas.length; j++) {
+            if (datas[j].refUID != bytes32(0)) {
+                canAttestToRef(datas[j].refUID);
+            }
+        }
+    }
+
+    /// 
+    /// Revokes multiple attestations
+    /// 
+    function multiRevoke(MultiRevocationRequest[] calldata multiRequests)
+        external
+        payable
+    {
+        // Checks if every revoke request belongs to the sender
+        // The sender can be either the attester or the recipient.
+        for (uint256 i = 0; i < multiRequests.length; i++) {
+            MultiRevocationRequest memory chunk = multiRequests[i];
+            for (uint256 j = 0; j < chunk.data.length; j++) {
+                Attestation memory target = eas.getAttestation(
+                    chunk.data[j].uid
+                );
+
+                require(
+                    target.attester == msg.sender ||
+                        target.recipient == msg.sender,
+                    /* || msg.sender == owner()*/
+                    // ↑ Should owner be able to revoke any attestation?
+                    "Not owner."
+                );
+            }
+        }
+        eas.multiRevoke(multiRequests);
+    }
+
+    function attest(AttestationRequest calldata request)
+        external
+        payable
+        returns (bytes32)
+    {
+        if (request.data.refUID != bytes32(0)) {
+            canAttestToRef(request.data.refUID);
+        }
+
+        return eas.attest(request);
+    }
+
+    function multiSequentialAttest(
+        AttestationRequestNode[] calldata requestNodes
+    ) public {
         bytes32[][] memory totalUids = new bytes32[][](requestNodes.length);
-        for(uint256 i = 0; i < requestNodes.length; i++) {
-            MultiAttestationRequest memory request = requestNodes[i].multiRequest;
+
+        for (uint256 i = 0; i < requestNodes.length; i++) {
+            MultiAttestationRequest memory request = requestNodes[i]
+                .multiRequest;
+            // If first item reference an attestation, checks if sender
+            // is owner or attester of that attestation.
+            canAttestToRefs(request.data);
+            // Updates the upcoming attestation reference uids.
             if (i > 0) {
-                for(uint256 j = 0; j < request.data.length; j++) {
+                for (uint256 j = 0; j < request.data.length; j++) {
                     AttestationRequestData memory data = request.data[j];
-                    data.refUID = totalUids[requestNodes[i].refIdx][0];
-                    request.data[j] = data;
+                    // If a request already has a ref, should not change it.
+                    if (data.refUID == bytes32(0)) {
```

Comment:
I can't move the ref logic to the validation but I actually can remove that call to canAttestToRef, because I'm checking it already in the line 100 :)